internal/servicelog/iterator.go

```diff
@@ -0,0 +1,175 @@
+// Copyright (c) 2021 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>
+
+package servicelog
+
+import (
+	"io"
+)
+
+type Iterator interface {
+	// Close closes the iterator so that buffers can be used for future writes.
+	// If Close is not called, the iterator will block buffer recycling causing
+	// write failures.
+	Close() error
+	// Next moves the ring buffer read mark forward, making its tail available for reuse
+	// without truncation. If the ring buffer writer produces data faster than the iterator
+	// can read it, the iterator will eventually be truncated and restarted. The truncation
+	// will be identified in the iterator output with the text specified when the iterator was
+	// created.
+	Next(cancel <-chan struct{}) bool
+
+	Buffered() int
+	io.Reader
+	io.WriterTo
+}
+
+type iterator struct {
+	rb           *RingBuffer
+	index        RingPos
+	trunc        []byte
+	truncWritten bool
+	notifyChan   chan bool
+	closeChan    chan struct{}
+}
+
+var _ Iterator = (*iterator)(nil)
+
+var (
+	truncBytes = []byte("(... output truncated ...)\n")
```

Comment:
I suppose this might end up at the end of a long line and be completely missed.

If so, let's add a "\n" as the first char as well to make sure the bump is noticeable on the output.

Reply:
Fixed.